### PR TITLE
`Tutorial groups`: Add the possibility for students and tutors to get notifications about tutorial groups

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The following notifications have been added:

|                                            | Type                          | Who                           | When?                                                          | E-Mail Support? |
|--------------------------------------------|-------------------------------|-------------------------------|----------------------------------------------------------------|-----------------|
| TUTORIAL_GROUP_REGISTRATION_STUDENT        | SingleUserNotification        | Student who got registered    | When registered to a tutorial group                            | Y               |
| TUTORIAL_GROUP_DEREGISTRATION_STUDENT      | SingleUserNotification        | Student who got deregistered  | When deregistered to a tutorial group                          | Y               |
| TUTORIAL_GROUP_REGISTRATION_TUTOR          | SingleUserNotification        | Tutor who leads group         | When student gets registered to group                          | Y               |
| TUTORIAL_GROUP_DEREGISTRATION_TUTOR        | SingleUserNotification        | Tutor who leads group         | When student gets deregistered from group                      | Y               |
| TUTORIAL_GROUP_REGISTRATION_MULTIPLE_TUTOR | SingleUserNotification        | Tutor who leads group         | When multiple students get registered in one go   (csv import) | Y               |
| TUTORIAL_GROUP_ASSIGNED                    | SingleUserNotification        | Tutor who leads group         | When tutor has been assigned to lead a group                   | Y               |
| TUTORIAL_GROUP_UNASSIGNED                  | SingleUserNotification        | Tutor who leads group         | When tutor has been deassigned from leading a group            | Y               |
| TUTORIAL_GROUP_DELETED                     | **TutorialGroupNotification** | Registered Students and Tutor | When tutorial group gets deleted                               | Y               |
| TUTORIAL_GROUP_UPDATED                     | **TutorialGroupNotification** | Registered Students and Tutor | When tutorial group gets updated  (for example title change)   | Y               |

### Description

As you can see in the table above, most of the new notifications have been implemented using the existing `SingleUserNotification` framework. But we need to also have a way to contact all registered students + the tutor of a tutorial group without sending a SingleUserNotification to each one independently. Imagine for example if a session of a tutorial group gets cancelled because of a holiday.

Therefore I implemented `TutorialGroupNotification` similar to `GroupNotification`. image

For each tutorial group in an active course where the user is registered or is the tutor, a websocket topic is created to which the respective notifications will be sent.

### Steps for Testing

- Activate E-Mail in the notification settings for each of the settings in the table above.
- Try to cause the notification to happen and check that the respective users have received a web notification + E-Mail. 
- ** Now deactivate the notifications (web app and email) in the settings and check that the notifications are not sent any more**

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line | Comment         |
|------------|-------:|-----:|-----------------|
| tutorial-groups-notification.service.ts| 80% | 100% | |
| notification.service.ts | 43% | 75 % | Added tests concerning my changes |

### Screenshots

**New Settings:**
image

### NOTE: I already fixed the problem with the Umlaute in the E-Mail but do not want to update all the screenshots

**Student Deregistration Notification and EMAIL**
image
image
**Student Registration Notification and EMAIL**
image
image
**Tutorial Group Update Notification and Email**
image
image
**Tutorial Group Unassigned Notification and Email**
image
image
**Tutorial Group Assigned Notification and Email**
image
image
* Tutorial Group Multiple Student Notification and Email**
image
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/notifications/NotificationSettingsService.java [hunk 1]
```diff
@@ -121,22 +145,17 @@ public NotificationSettingsService(NotificationSettingRepository notificationSet
...
-            deactivatedTypes = findDeactivatedNotificationTypes(communicationChannel, DEFAULT_NOTIFICATION_SETTINGS);
-        }
-        else {
-            deactivatedTypes = findDeactivatedNotificationTypes(communicationChannel, notificationSettings);
-        }
-
-        if (deactivatedTypes.isEmpty()) {
-            return true;
+        // for those notification types that are not explicitly set by the user, we use the default settings
+        Set<String> decidedIds = decidedNotificationSettings.stream().map(NotificationSetting::getSettingId).collect(Collectors.toSet());
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/service/notifications/NotificationSettingsService.java:152]: This fixes a bug that I noticed. When a new notification setting has been introduced, there is no corresponding row in the notification settings table. This caused e-mail to always be sent even though the default settings actually specify no email.

**Comment by stefanwaldhauser** [comment]
Fixing the god class codacy issue is out of scope of this PR imo: 
image

File: src/main/java/de/tum/in/www1/artemis/service/notifications/TutorialGroupNotificationService.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    private void saveAndSend(TutorialGroupNotification notification, boolean notifyTutor) {
+        tutorialGroupNotificationRepository.save(notification);
+        sendNotificationViaWebSocket(notification);
+        sendNotificationViaMail(notification, notifyTutor);
+    }
+
+    private void sendNotificationViaMail(TutorialGroupNotification notification, boolean notifyTutor) {
+        if (notificationSettingsService.checkNotificationTypeForEmailSupport(notification.notificationType)) {
+            var usersToMail = findUsersToMail(notification, notifyTutor);
+            if (!usersToMail.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/notifications/TutorialGroupNotificationService.java:77]: `usersToMail` is a Set and not an Optional

File: src/main/java/de/tum/in/www1/artemis/domain/notification/SingleUserNotificationFactory.java [hunk 3]
```diff
@@ -115,4 +121,86 @@ public static SingleUserNotification createNotification(PlagiarismCase plagiaris
...
+        SingleUserNotification notification;
+        switch (notificationType) {
+            case TUTORIAL_GROUP_REGISTRATION_STUDENT -> {
+                var student = users.stream().findAny().orElseThrow();
+                notification = new SingleUserNotification(student, title,
+                        "You have been registered to the tutorial group " + tutorialGroup.getTitle() + " by " + responsibleForAction.getName() + ".");
+                notification.setTransientAndStringTarget(createTutorialGroupTarget(tutorialGroup, tutorialGroup.getCourse().getId(), false, true));
+            }
+            case TUTORIAL_GROUP_DEREGISTRATION_STUDENT -> {
+
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/notification/SingleUserNotificationFactory.java]: Removed it in 25d612935c597106746b13b4799158801e534b3d

File: src/main/java/de/tum/in/www1/artemis/repository/tutorialgroups/TutorialGroupRepository.java [hunk 4]
```diff
@@ -60,6 +63,8 @@ public interface TutorialGroupRepository extends JpaRepository<TutorialGroup, Lo
 
     Set<TutorialGroup> findAllByTeachingAssistant(User teachingAssistant);
 
+    @Transactional
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/repository/tutorialgroups/TutorialGroupRepository.java]: Added a comment in 25d612935c597106746b13b4799158801e534b3d

File: src/main/java/de/tum/in/www1/artemis/service/MailService.java [hunk 5]
```diff
@@ -207,6 +215,10 @@ public void sendNotificationEmail(Notification notification, User user, Object n
             context.setVariable(PLAGIARISM_VERDICT, plagiarismCase.getVerdict());
         }
 
+        if (notificationSubject instanceof SingleUserNotificationService.TutorialGroupNotificationSubject) {
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/MailService.java]: Wow did not know about this. Seems to be introduced in Java 14. Used it in 25d612935c597106746b13b4799158801e534b3d

File: src/main/webapp/i18n/de/userSettings.json [hunk 6]
```diff
@@ -22,7 +22,9 @@
                 "weeklySummary": "Wöchentliche Zusammenfassung",
                 "exerciseNotifications": "Aufgabenbenachrichtigungen",
                 "lectureNotifications": "Vorlesungsbenachrichtigungen",
+                "tutorialGroupNotifications": "Übungsbereichsbenachrichtigungen",
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/webapp/i18n/de/userSettings.json]: Changed it in 25d612935c597106746b13b4799158801e534b3d

File: src/main/webapp/i18n/de/userSettings.json [hunk 7]
```diff
@@ -58,7 +63,11 @@
                 "newAnnouncementPostDescription": "Erhalte eine Mitteilung, wenn eine neue Ankündigung erstellt wurde",
                 "courseAndExamArchivingStartedDescription": "Erhalte eine Mitteilung, wenn das Archivieren eines Kurses oder einer Klausur angefangen hat",
                 "fileSubmissionSuccessfulDescription": "Erhalte eine Mitteilung nach dem erfolgreichen Einreichen von Dateiupload-Aufgaben",
-                "programmingTestCasesChangedDescription": "Erhalte eine Mitteilung bei Änderungen von Testfällen in Programmieraufgaben"
+                "programmingTestCasesChangedDescription": "Erhalte eine Mitteilung bei Änderungen von Testfällen in Programmieraufgaben",
+                "registrationTutorialGroupTutorDescription": "Erhalte eine Mitteilung, wenn ein Student für deine Übungsgruppe registriert oder abgemeldet wurde",
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/webapp/i18n/de/userSettings.json]: Good catch! Changed it in 25d612935c597106746b13b4799158801e534b3d

**Review by JohannesStoehr — APPROVED** [review]
Code looks good to me after fixes

File: src/main/java/de/tum/in/www1/artemis/service/notifications/TutorialGroupNotificationService.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    /**
+     * Notify registered students about an updated tutorial group
+     *
+     * @param tutorialGroup the updated tutorial group
+     * @param notifyTutor   whether the tutor should be notified about the update via email
+     */
+    public void notifyAboutTutorialGroupUpdate(TutorialGroup tutorialGroup, boolean notifyTutor) {
+        saveAndSend(createTutorialGroupNotification(tutorialGroup, TUTORIAL_GROUP_UPDATED), notifyTutor);
```
Comment (ge65cer) [inline src/main/java/de/tum/in/www1/artemis/service/notifications/TutorialGroupNotificationService.java:65]: Must: Do not notify the user about changes to the tutorial group they initiated themselves.

Optional: I do not think this notification makes sense from a user perspective right now. It should either include what got updated (only important changes such as time or place), or be removed.

File: src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java [hunk 9]
```diff
@@ -175,38 +208,59 @@ public ResponseEntity<Void> delete(@PathVariable Long courseId, @PathVariable Lo
...
+        // Note: We have to load the teaching assistants from database otherwise languageKey is not defined and email sending fails
+        if (oldTutorialGroup.getTeachingAssistant() == null && updatedTutorialGroup.getTeachingAssistant() != null) {
+            var newTAFromDatabase = userRepository.findOneByLogin(updatedTutorialGroup.getTeachingAssistant().getLogin());
+            newTAFromDatabase.ifPresent(user -> singleUserNotificationService.notifyTutorAboutAssignmentToTutorialGroup(updatedTutorialGroup, user, responsibleUser));
+        }
+        else if (oldTutorialGroup.getTeachingAssistant() != null && updatedTutorialGroup.getTeachingAssistant() == null) {
+            var oldTAFromDatabase = userRepository.findOneByLogin(oldTutorialGroup.getTeachingAssistant().getLogin());
+            oldTAFromDatabase.ifPresent(user -> singleUserNotificationService.notifyTutorAboutUnassignmentFromTutorialGroup(oldTutorialGroup, user, responsibleUser));
+        }
+        else if (oldTutorialGroup.getTeachingAssistant() != null && updatedTutorialGroup.getTeachingAssistant() != null
```
Comment (ge65cer) [inline src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:252]: Why do you need an extra case? Couldn't you cover this case in two if statements?
```
Case 1: new != null && (old == null || old != new) → send assignment notification to new
Case 2: old != null && (new == null || old != new) → send unassignment notification to old
```

File: src/main/resources/i18n/messages_de.properties [hunk 10]
```diff
@@ -98,5 +98,16 @@ email.notification.title.post.plagiarismVerdict=Entscheidung zum Plagiatsfall in
 email.notification.title.post.plagiarismCase=Neuer Plagiatsfall in Aufgabe {0}
 email.notification.aux.plagiarismVerdict.plagiarism=Der Fall wird als Plagiat angesehen!
 email.notification.aux.plagiarismVerdict.point.deduction=Wegen des Plagiatsfalls ziehen wir dir Punkte in der Aufgabe ab!
-email.notification.aux.plagiarismVerdict.warning=Du wirst wegen des Plagiatsversuchs verwarnt. Weitere Plagiate k�nnen schwere Folgen haben!
+email.notification.aux.plagiarismVerdict.warning=Du wirst wegen des Plagiatsversuchs verwarnt. Weitere Plagiate können schwere Folgen haben!
 email.notification.aux.plagiarismVerdict.no.plagiarism=Der Fall wird nicht mehr als Plagiat angesehen.
+
+# Tutorial Groups
+email.notification.title.tutorialGroup.registrationStudent=Du wurdest für die Übungsgruppe {0} im Kurs {1} registriert von {2}.
```
Comment (ge65cer) [inline src/main/resources/i18n/messages_de.properties:105]: See English messages file.

File: src/main/resources/i18n/messages_en.properties [hunk 11]
```diff
@@ -100,3 +100,15 @@ email.notification.aux.plagiarismVerdict.plagiarism=Your case is considered as p
 email.notification.aux.plagiarismVerdict.point.deduction=Due to your plagiarism, we are deducting points from your score in this exercise!
 email.notification.aux.plagiarismVerdict.warning=Due to the plagiarism attempt, you are warned. Any further plagiarism can lead to severe consequences!
 email.notification.aux.plagiarismVerdict.no.plagiarism=The case is no longer considered as plagiarism.
+
+# Tutorial Groups
+email.notification.title.tutorialGroup.registrationStudent=You have been registered to the tutorial group {0} in the course {1} by {2}.
```
Comment (ge65cer) [inline src/main/resources/i18n/messages_en.properties:105]: In my opinion active would sound a little better, e.g., “{2} registered you to the tutorial group {0} in the course {1}“.

File: src/main/java/de/tum/in/www1/artemis/domain/notification/SingleUserNotificationFactory.java [hunk 12]
```diff
@@ -115,4 +119,82 @@ public static SingleUserNotification createNotification(PlagiarismCase plagiaris
...
+                var tutorToContact = users.stream().findAny().get();
+                notification = new SingleUserNotification(tutorToContact, title,
+                        "You have been assigned to lead the tutorial group " + tutorialGroup.getTitle() + " by " + responsibleForAction.getName() + ".");
+                notification.setTransientAndStringTarget(createTutorialGroupTarget(tutorialGroup, tutorialGroup.getCourse().getId(), true, true));
+            }
+            case TUTORIAL_GROUP_UNASSIGNED -> {
+                var tutorToContact = users.stream().findAny().get();
+                notification = new SingleUserNotification(tutorToContact, title,
+                        "You have been unassigned from leading the tutorial group " + tutorialGroup.getTitle() + " by " + responsibleForAction.getName() + ".");
+                notification.setTransientAndStringTarget(createTutorialGroupTarget(tutorialGroup, tutorialGroup.getCourse().getId(), true, true));
```
Comment (ge65cer) [inline src/main/java/de/tum/in/www1/artemis/domain/notification/SingleUserNotificationFactory.java:193]: Q: This notification currently creates a link to `https://artemis-test1.artemis.in.tum.de/course-management/24/tutorial-groups-management/18`, which might not make sense when I was unassigned from leading the group (do I still have permission as normal tutor)?

**Review by ge65cer — CHANGES_REQUESTED** [review]
Tested on TS1 and successfully received all e-mails. 👍 
Left some comments below.

File: src/main/java/de/tum/in/www1/artemis/service/notifications/TutorialGroupNotificationService.java (hunk 8 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/notifications/TutorialGroupNotificationService.java:65]: I solved it now as we have it with exercises. The user can enter an optional notification text in the update form, if he/she wants a notification to be sent: 
image

Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/notifications/TutorialGroupNotificationService.java]: Did it in 1ec6e2b257b4d09878415389993838bba7e417ea

File: src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java (hunk 9 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java]: I like it! Much cleaner :) Thank you! Changed in 00f5ce4fc7e8717bcec8b07181762180e3041dd5

File: src/main/java/de/tum/in/www1/artemis/domain/notification/SingleUserNotificationFactory.java (hunk 12 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/notification/SingleUserNotificationFactory.java:193]: Yeah. You can still see the detail page. But the edit buttons and such will be deactivated

File: src/main/resources/i18n/messages_en.properties (hunk 11 above)
Comment (stefanwaldhauser) [reply src/main/resources/i18n/messages_en.properties]: Changed to active form in a2d98006db19207c790b367b9e6479bd55ba6c9d

File: src/main/resources/i18n/messages_de.properties (hunk 10 above)
Comment (stefanwaldhauser) [reply src/main/resources/i18n/messages_de.properties]: Changed to active form in a2d98006db19207c790b367b9e6479bd55ba6c9d

File: src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java [hunk 13]
```diff
@@ -175,38 +211,71 @@ public ResponseEntity<Void> delete(@PathVariable Long courseId, @PathVariable Lo
...
+
     /**
      * PUT /courses/:courseId/tutorial-groups/:tutorialGroupId : Updates an existing tutorial group
      *
-     * @param courseId        the id of the course to which the tutorial group belongs to
-     * @param tutorialGroupId the id of the tutorial group to update
-     * @param tutorialGroup   group the tutorial group to update
+     * @param courseId               the id of the course to which the tutorial group belongs to
+     * @param tutorialGroupId        the id of the tutorial group to update
+     * @param tutorialGroupUpdateDTO dto containing the t utorial group to update and the optional notification text
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:234]:
```suggestion
     * @param tutorialGroupUpdateDTO dto containing the tutorial group to update and the optional notification text
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java [hunk 14]
```diff
@@ -175,38 +211,71 @@ public ResponseEntity<Void> delete(@PathVariable Long courseId, @PathVariable Lo
...
-        trimStringFields(tutorialGroup);
-        if (!tutorialGroupFromDatabase.getTitle().equals(tutorialGroup.getTitle())) {
-            checkTitleIsUnique(tutorialGroup);
+        // Note: We have to load the teaching assistants from database otherwise languageKey is not defined and email sending fails
+
+        var oldTA = oldTutorialGroup.getTeachingAssistant();
+        var newTA = updatedTutorialGroup.getTeachingAssistant();
+
+        if (newTA != null && (oldTA == null || !oldTA.equals(newTA))) {
+            var newTAFromDatabase = userRepository.findOneByLogin(updatedTutorialGroup.getTeachingAssistant().getLogin());
```
Comment (ge65cer) [inline src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:263]:
```suggestion
            var newTAFromDatabase = userRepository.findOneByLogin(newTA.getLogin());
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java [hunk 15]
```diff
@@ -175,38 +211,71 @@ public ResponseEntity<Void> delete(@PathVariable Long courseId, @PathVariable Lo
...
+            newTAFromDatabase.ifPresent(user -> singleUserNotificationService.notifyTutorAboutAssignmentToTutorialGroup(updatedTutorialGroup, user, responsibleUser));
+        }
+        if (oldTA != null && (newTA == null || !newTA.equals(oldTA))) {
+            var oldTAFromDatabase = userRepository.findOneByLogin(oldTutorialGroup.getTeachingAssistant().getLogin());
```
Comment (ge65cer) [inline src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:267]:
```suggestion
            var oldTAFromDatabase = userRepository.findOneByLogin(oldTA.getLogin());
```

File: src/main/java/de/tum/in/www1/artemis/service/notifications/TutorialGroupNotificationService.java (hunk 8 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/notifications/TutorialGroupNotificationService.java]: image

**Review by ge65cer — APPROVED** [review]
My previous concerns were addressed. 💯

**Review by JohannesStoehr — APPROVED** [review]
Rechecked the latest commits. Code looks good to me

**Pull request merged**